src/MelodyParser.js

```diff
@@ -0,0 +1,173 @@
+const durations = require('./durations.js');
+const note = require('./note.js');
+
+class MelodyParser {
+
+    //interface method: song - text song, melody - MelodyQueue
+    parse(song, melody) {
+
+        //reset melody queue
+        melody.queue = [];
+
+        this.melody = melody;
+
+        //parse to melody
+        let lines = this.lines(song);
+        console.log('Preparing.');
+        this.prepareCommands(lines);
+        console.log('Building melody...');
+
+        melody = this.buildMelody();
+        console.log('Parsing complete...');
+
+    }
+
+    //split to lines and cleanup - rid of unwanted spaces
+    lines(sng) {
+        //reset parser
+        this.defines = [];
+        this.plays = [];
+        this.sounds = [];
+
+        //initial cleaning
+
+        // make single space separator
+        while (sng !== (sng = sng.replace(/  /g, ' '))) {};
+
+        sng = sng.replace(/\( /g, '(');
+        sng = sng.replace(/ \(/g, '(');
+        sng = sng.replace(/ \)/g, ')');
+        sng = sng.replace(/ ,/g, ',');
+        sng = sng.replace(/, /g, ',');
+        sng = sng.replace(/ ;/g, ';');
+        sng = sng.replace(/; /g, ';');
+        sng = sng.split("\n");
+        console.log(sng);
+        //make duration separator ready for split, return cleaned song definition
+        return sng;
+    }
+
+    prepareCommands(lins) {
+
+        let comm = [];
+
+        // as 'define' not always occurs as first, sort at the begining would be gr8
+        // but warn information wouldn't be working properly.
+
+        let plays = []; // let's store at first pass
+        let cLean = '';
+        console.log('Parsing defines...');
+
+        for (let l = 0; l < lins.length; l++) {
+
+            // additional cleaning
+            lins[l] = lins[l].trim();
+
+            //separate commands
+            comm = lins[l].split(' ');
+
+            switch (comm[0]) {
+                case 'define':
+                    this.defines = Object.assign(this.defines, this.parseCommandDefine(comm));
+                    break;
+
+                case 'play':
+                    plays.push(comm);
+                    break;
+
+                default:
+                    console.log(`WARN: Command not recognized or empty line (${l+1}). `)
+
+                    break;
+            }
+        }
+        console.log('Parsing notes...');
+        for (let p = 0; p < plays.length; p++) {
+            this.parseCommandPlay(plays[p]);
+        }
+    }
+
+    //parse definitions (aka macros)
+    parseCommandDefine(comm) {
+        let defs = [];
+
+        let name = comm[1];
+        for (let i = 3; i < comm.length; i++) {
+            if (comm[i] !== '|') { // ignore pipe separator
```

Comment:
better is ifError pattern:
```js
if (comm[i] === '|') {
  continue; // or return if is a function
}```

Reply:
Better because?

Reply:
1. Code is more flat (not nested)
2. Avoid this: https://external-preview.redd.it/I1CShqp2wBguKcuv21m6XRnREKuw8kyAoCavCAGJPBI.jpg?auto=webp&s=819f73b35497078f717311ec509e8224b89b6e71

See: http://wiki.c2.com/?IfOk

Reply:
Good point :)